typing/env.ml

```diff
@@ -3584,93 +3584,98 @@ let extract_instance_variables env =
        | Val_ivar _ -> name :: acc
        | _ -> acc) None env []
 
-let report_lookup_error_doc _loc env ppf = function
-  | Unbound_value(lid, hint) -> begin
-      fprintf ppf "Unbound value %a" quoted_longident lid;
-      spellcheck ppf extract_values env lid;
-      match hint with
-      | No_hint -> ()
-      | Missing_rec def_loc ->
+let report_lookup_error_doc loc env = function
+  | Unbound_value(lid, hint) ->
+     let spellcheck_hint = spellcheck extract_values env lid in
+     let rec_hint = match hint with
+       | No_hint -> []
+       | Missing_rec def_loc ->
           let (_, line, _) =
             Location.get_pos_info def_loc.Location.loc_start
           in
-          fprintf ppf
-            "@.@[@{<hint>Hint@}: If this is a recursive definition,@ \
-             you should add the %a keyword on line %i@]"
-            Style.inline_code "rec"
-            line
-    end
+          [Location.msg
+             "@[@{<hint>Hint@}: If this is a recursive definition,@ \
+              you should add the %a keyword on line %i@]"
+             Style.inline_code "rec"
+             line]
+     in
+     let sub = spellcheck_hint @ rec_hint in
+     Location.errorf ~loc ~sub "Unbound value %a" quoted_longident lid
```

Comment:
Here and in many other instances of the refactoring that you did, I'm not fond of the fact that:
- before, the code read in the order of the messages being printed: first the main error message, then the hint that is printed
- after the refactoring, we get to see the hint first and the code later

This is a bit unfortunate because often the error constructor names are not completely clear, and reading the error message helps knowing what this case is about. Swapping the order of the error message bits makes this code a bit harder to read, in my opinion.

Could you use a cheap trick to restore reading order for the literal strings? Maybe something like

```ocaml
let msg ~sub = 
  ...
in
let sub =
  ...
in
msg sub
```

Reply:
While rebasing over the alignment change, I have restored the correct order between the main error message and the secondary ones.